llumnix/backends/vllm/migration_backend.py

```diff
@@ -0,0 +1,279 @@
+            offset = min(self.num_migration_cache_blocks, tot_blocks - start_idx)
+            send_blocks = src_blocks[start_idx:start_idx+offset]
+            ray_obj = self.actor.exec_method.remote(self.is_driver_worker, src_handle, "do_send", send_blocks)
+            if rpc_numpy_cache is not None:
+                self.do_recv(rpc_numpy_cache, recv_blocks)
+            rpc_numpy_cache = ray.get(ray_obj)
+            recv_blocks = dst_blocks[start_idx:start_idx+offset]
+        self.do_recv(rpc_numpy_cache, recv_blocks)
+
+    def do_send(self, blocks: List[int]):
+        num_blocks = len(blocks)
+        send_cache = self.dummy_cache[:num_blocks].view(self.num_layers, 2, num_blocks, self.migration_cache_size)
+        src_to_dst = {block_num: idx for idx, block_num in enumerate(blocks)}
+        with torch.cuda.stream(self.migration_stream):
+            for layer_idx in range(self.num_layers):
+                self.cache_engine.attn_backend.swap_blocks(self.gpu_cache[layer_idx], send_cache[layer_idx], src_to_dst)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+        return send_cache.to(self.rpc_dtype).numpy()
+
+    def do_recv(self, rpc_numpy_cache, blocks: List[int]):
+        num_blocks = len(blocks)
+        src_to_dst = dict(enumerate(blocks))
+        recv_cache = self.dummy_cache[:num_blocks].view(self.num_layers, 2, num_blocks, self.migration_cache_size)
+        # use pin memory dummy_cache to speed up data transfer
+        recv_cache.copy_(torch.from_numpy(rpc_numpy_cache))
+
+        with torch.cuda.stream(self.migration_stream):
+            for layer_idx in range(self.num_layers):
+                self.cache_engine.attn_backend.swap_blocks(recv_cache[layer_idx], self.gpu_cache[layer_idx], src_to_dst)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+
+class RayColMigrationBackend(MigrationBackendBase):
+    def __init__(self, migration_config: MigrationConfig, cache_engine: CacheEngine, local_rank,
+                 scheduling_strategy, is_driver_worker, gpu_cache) -> None:
+        super().__init__()
+
+        self.migration_config = migration_config
+        self.cache_engine = cache_engine
+        self.backend = migration_config.migration_backend
+        self.num_migration_num_layers = min(migration_config.migration_num_layers, self.cache_engine.num_layers)
+        self.num_migration_cache_blocks = migration_config.migration_cache_blocks
+
+        self.backend = migration_config.migration_backend
+        self.ray_world_size = -1
+        self.ray_rank = -1
+        self.group_name = None
+
+        self.local_rank = local_rank
+        self.actor = ProxyActor.options(scheduling_strategy=scheduling_strategy).remote()
+        self.is_driver_worker = is_driver_worker
+        self.gpu_cache = gpu_cache
+        # add bf16 type
+        gloo_util.TORCH_GLOO_DTYPE_MAP[torch.bfloat16] = pygloo.glooDataType_t.glooFloat16
```

Comment:
add comment:
Add support for bf16 type in Gloo. Now bf16 and fp16 both map to glooFloat16, but this is okay because Gloo only uses the data type size for transmission.